Lib/asyncio/runners.py

```diff
@@ -1,10 +1,85 @@
-__all__ = 'run',
+__all__ = ('Runner', 'run')
 
+import contextvars
 from . import coroutines
 from . import events
 from . import tasks
 
 
+class Runner:
+    """A context manager that controls event loop life cycle.
+
+    The context manager always creates a new event loop,
+    allows to run async functions inside it,
+    and properly finalizes the loop at the context manager exit.
+
+    If debug is True, the event loop will be run in debug mode.
+    If factory is passed, it is used for new event loop creation.
+
+    asyncio.run(main(), debug=True)
+
+    is a shortcut for
+
+    with asyncio.Runner(debug=True) as runner:
+        runner.run(main())
+
+    The run() method can be called multiple times within the runner's context.
+
+    This can be useful for interactive console (e.g. IPython),
+    unittest runners, console tools, -- everywhere when async code
+    is called from existing sync framework and where the preferred single
+    asyncio.run() call doesn't work.
+
+    """
+    def __init__(self, *, debug=None, factory=None):
+        if factory is None:
+            self._loop = events.new_event_loop()
```

Comment:
The original `asyncio.run()` implementation calls `set_event_loop()`. I think Runner should also call it.

Reply:
`set_event_loop` is not necessary inside of `run_until_complete()` called coroutine.
I can rather suggest a deprecation of `set_event_loop()` and using the loop instance explicitly if the loop is not run.
Inside a running loop the loop is globally accessible anyway.

Reply:
Since Python 3.10 `get_event_loop()` raises DeprecationWarning if used outside of running loop.
If it is called from the loop context it works as an alias for `get_running_loop()`.
Smart and backward compatible, thanks to @serhiy-storchaka 

`set_event_loop()` is another beast. I support its deprecation. Now it is useless unless `get_event_loop()` is called in deprecated mode.

Reply:
`set_event_loop()` deprecation will be done in another PR.